File: scripts/seed-users.ts
/**
 * Seed 30 users into Supabase Auth:
 *   2 admin, 8 manager, 20 readonly
 *
 * Usage: npm run seed-users
 * Requires: SUPABASE_SERVICE_ROLE_KEY + NEXT_PUBLIC_SUPABASE_URL in .env.local
 *
 * Credentials are printed to stdout — save them somewhere secure.
 * This script is idempotent: re-running skips existing users.
 */

import { createClient } from "@supabase/supabase-js";
import { readFileSync } from "fs";
import { resolve } from "path";

// ── Load .env.local (tsx --env-file handles this at runtime, but parse manually as fallback) ──
function loadEnvLocal(): void {
  try {
    const envPath = resolve(process.cwd(), ".env.local");
    const content = readFileSync(envPath, "utf-8");
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) continue;
      const eq = trimmed.indexOf("=");
      if (eq < 0) continue;
      const key = trimmed.slice(0, eq).trim();
      const value = trimmed.slice(eq + 1).trim();
      if (!process.env[key]) process.env[key] = value;
    }
  } catch {
    // .env.local not found — rely on process.env (e.g. CI vars or tsx --env-file)
  }
}

loadEnvLocal();

const URL = process.env.NEXT_PUBLIC_SUPABASE_URL!;
const KEY = process.env.SUPABASE_SERVICE_ROLE_KEY!;

if (!URL || !KEY) {
  console.error(
    "Missing env vars. Set NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY in .env.local",
  );
  process.exit(1);
}

const supabase = createClient(URL, KEY, {
  auth: { autoRefreshToken: false, persistSession: false },
});

interface SeedUser {
  email: string;
  password: string;
  role: "admin" | "manager" | "readonly";
}

const PASSWORD = "mwlab2026!";

function buildUserList(): SeedUser[] {
  const users: SeedUser[] = [];

  // 2 admin
  for (let i = 0; i < 2; i++) {
    users.push({
      email: `user${String(i).padStart(2, "0")}@mwlab.internal`,
      password: PASSWORD,
      role: "admin",
    });
  }

  // 8 manager
  for (let i = 2; i < 10; i++) {
    users.push({
      email: `user${String(i).padStart(2, "0")}@mwlab.internal`,
      password: PASSWORD,
      role: "manager",
    });
  }

  // 20 readonly
  for (let i = 10; i < 30; i++) {
    users.push({
      email: `user${String(i).padStart(2, "0")}@mwlab.internal`,
      password: PASSWORD,
      role: "readonly",
    });
  }

  return users;
}

async function main(): Promise<void> {
  const users = buildUserList();
  const credentials: { email: string; password: string; role: string }[] = [];

  console.log(`Seeding ${users.length} users...\n`);

  for (const u of users) {
    // Check if user already exists (idempotency)
    const { data: existing } = await supabase.auth.admin.listUsers();
    const alreadyExists = existing?.users?.some(
      (eu) => eu.email === u.email,
    );

    if (alreadyExists) {
      console.log(`  SKIP ${u.email} — already exists`);
      continue;
    }

    const { data, error } = await supabase.auth.admin.createUser({
      email: u.email,
      password: u.password,
      email_confirm: true,
      app_metadata: { role: u.role },
    });

    if (error) {
      console.error(`  FAIL ${u.email}: ${error.message}`);
      continue;
    }

    console.log(`  OK   ${u.email} (${u.role})`);
    credentials.push({ email: u.email, password: u.password, role: u.role });
  }

  // ── Summary ──
  console.log("\n─── Seed complete ───");
  console.log(
    `Created: ${credentials.length} | Skipped: ${users.length - credentials.length}`,
  );

  // Print admin credentials separately
  const admins = credentials.filter((c) => c.role === "admin");
  if (admins.length > 0) {
    console.log("\nAdmin logins:");
    for (const a of admins) {
      console.log(`  ${a.email} / ${a.password}`);
    }
  }
}

main().catch((err) => {
  console.error("Seed failed:", err);
  process.exit(1);
});

File: scripts/seed_users.py
"""
seed_users.py — 向 data/mwlab.db 的 user 表插入初始开发账号（bcrypt 密码哈希）。

用法:
    python3 scripts/seed_users.py

幂等：已存在的 email 会跳过。

⚠️ 下列为本地开发用弱口令，生产环境务必改密（见 README「用户与权限」）。
"""
from __future__ import annotations

import sqlite3
from pathlib import Path

import bcrypt

DB_PATH = Path(__file__).resolve().parent.parent / "data" / "mwlab.db"

USERS = [
    {
        "email": "admin@mwlab.internal",
        "password": "admin123",
        "role": "admin",
        "display_name": "管理员",
    },
    {
        "email": "manager@mwlab.internal",
        "password": "manager123",
        "role": "manager",
        "display_name": "经理",
    },
    {
        "email": "readonly@mwlab.internal",
        "password": "readonly123",
        "role": "readonly",
        "display_name": "只读用户",
    },
]


def main() -> None:
    conn = sqlite3.connect(str(DB_PATH))
    try:
        existing = set(
            row[0]
            for row in conn.execute("SELECT email FROM user").fetchall()
        )

        inserted = 0
        skipped = 0
        for u in USERS:
            if u["email"] in existing:
                print(f"  SKIP  {u['email']} — 已存在")
                skipped += 1
                continue

            password_hash = bcrypt.hashpw(
                u["password"].encode("utf-8"), bcrypt.gensalt()
            ).decode("utf-8")
            conn.execute(
                """
                INSERT INTO user (email, password_hash, role, display_name, is_active)
                VALUES (?, ?, ?, ?, 1)
                """,
                (u["email"], password_hash, u["role"], u["display_name"]),
            )
            print(f"  OK    {u['email']} (role={u['role']})")
            inserted += 1

        conn.commit()
        print(f"\n完成：插入 {inserted}，跳过 {skipped}")
    finally:
        conn.close()


if __name__ == "__main__":
    main()
